parser/parser.go

```diff
@@ -0,0 +1,599 @@
+
+	var newtag Tag
+	var key string
+	var vendor string
+
+	p.Consume()
+	key = parseUntil(p, isHostnameRune)
+
+	r, ok := p.Next()
+	if !ok {
+		p.sendError("parser: (tag state) unexpected end of input")
+		p.Consume()
+		return nil
+	}
+
+	if r == '/' {
+		vendor = key
+		p.Consume()
+		key = parseUntil(p, isHostnameRune)
+		if len(key) == 0 {
+			p.sendError("parser: (tag state) missing hostname after vendor")
+			return errorState
+		}
+		r, ok = p.Next()
+		if !ok {
+			p.sendError("parser: (tag state) unexpected end of input")
+			p.Consume()
+			return nil
+		}
+	}
+	if strings.ContainsRune(key, '.') {
+		p.sendError("parser: (tag state) unexpected '.' in key")
+		return errorState
+	}
+	if vendor != "" {
+		if vendor[0] == '.' || vendor[0] == '-' {
+			p.sendError("parser: (tag state) unexpected '" + string(vendor[0]) + "' at beginning of vendor")
+			return errorState
+		}
+		if vendor[len(vendor)-1] == '.' {
+			vendor = vendor[:len(vendor)-1]
+		}
+		if len(vendor) > 253 {
+			p.sendError("parser: (tag state) hostname exceeds length limit of 253")
+			return errorState
+		}
+		for _, lbl := range strings.Split(vendor, ".") {
+			if len(lbl) <= 0 || len(lbl) > 63 {
+				p.sendError("parser: (tag state) hostname label must be between 1 and 63 characters long")
+				return errorState
+			}
+		}
+	}
+	if key == "" {
+		p.sendError("parser: (tag state) missing valid character after tag symbol '@'")
+		return errorState
+	}
+	newtag.Key = key
+	newtag.Vendor = vendor
+
+	// parse optional tag value
+	var value string
+	if r == '=' {
+		p.Consume()
+		// parse any number of runes except NUL, BELL, CR, LF, ';', ' '
+		value = parseUntil(p, isValueRune)
+		r, ok = p.Next()
+		if !ok {
+			p.sendError("parser: (tag state) unexpected end of input")
+			p.Consume()
+			return nil
+		}
+	}
+	newtag.Value = value
+
+	// parse optional ';' rune
+	if r == ';' {
+		// consume runes and store in tag list
+		p.msg.Tags = append(p.msg.Tags, newtag)
+		return tagState
+	}
+
+	// ending rune for all tags
+	if r != ' ' {
+		p.sendError("parser: (tag state) expected ' ' at end of tags")
+		return errorState
+	}
+	if !skipSpaces(p) {
+		p.sendError("parser: (tag state) unexpected end of input")
+		p.Consume()
+		return nil
+	}
+
+	// consume runes and store in tag list
+	p.msg.Tags = append(p.msg.Tags, newtag)
+	r, ok = p.Next()
+	if !ok {
+		p.sendError("parser: (tag state) unexpected end of input")
+		p.Consume()
+		return nil
+	}
+	// transition to new state
+	if r == ':' {
+		return prefixState
+	}
+
+	p.Rewind()
+	return commandState
+}
+
+func prefixState(p *Parser) StateFn {
+	p.Consume()
+	p.msg.Prefix.Name = parseUntil(p, isPrefixNameRune)
+
+	r, ok := p.Next()
+	if !ok {
+		p.sendError("parser: (prefix state) unexpected end of input")
+		p.Consume()
+		return nil
+	}
+
+	if r == '!' || r == '@' {
+		if r == '!' {
+			p.Consume()
+			p.msg.Prefix.Username = parseUntil(p, func(r rune) bool {
+				return r != '@'
```

Comment:
Consider `thatserver/johnny!johnnyman sometext@thathost`.
The username would be parsed as `johnnyman sometext`.
Are spaces allowed for usernames?

Reply:
Done. There is no concrete specification, so I made one up based on what I could test on some random IRC server.